### ISSUE TITLE
Change the new operator to use CREATE2

### ISSUE THREAD

**Issue by chriseth** [ethereum/solidity#2136]
This might require keeping a nonce or other measure to make the creation bytecode unique if `new` is used multpile times on the same contract.

See https://eips.ethereum.org/EIPS/eip-1014

**Comment by VoR0220** [comment]
Can you elaborate on this?

**Comment by axic** [comment]
Would that counter be shared across the contract for all creates, specific to the create in a certain code location or specific to creating a specific contract?

**Comment by AlexeyAkhunov** [comment]
I have written up this: https://ethereum-magicians.org/t/solidity-compiler-magicians-wanted-to-make-create2-usage-easier/2137

Yesterday I made this ERC20 contract (which is supposed to work under State rent when holders pay for storing their tokens): work: https://github.com/ledgerwatch/eth_state/blob/master/erc20/TokenContract.sol

As you can see in the code, Solidity provides ``create2`` assembly function, but working with it is cumbersome. It would be much easier if there were 2 new features in Solidity:

1) ``new2`` operator. Similar to ``new``, but also accept ``salt`` parameter, and uses ``CREARE`` opcode. So in my ``Factory`` I would have written ``Holder holder = new2(Holder, _owner)`` without having to put the byte code of the ``Holder`` contract inside the function
2) Function ``create2_address``, which allows computing address of ``CREATE2`` contract. I would have used it twice, first in ``Token.getHolderContract`` function: ``address payable holder_address = create2_address(factory, a, Holder)``, and second in ``Holder.setOwner`` function: ``require(create2_address(factor, _owner, Holder) == address(this))``

It would be great to have (at least in a branch) a version of Solidity (derived from the current) supporting these two things. That would make further development of contracts like that much easier.

**Comment by chriseth** [comment]
The new2 function you suggest could be rather tricky to implement. What might be easier and could be almost as helpful to you would be something like `Holder.creationCode` and `Holder.runtimeCode` - read-only properties of type `bytes memory`. What do you think?

**Comment by AlexeyAkhunov** [comment]
> What might be easier and could be almost as helpful to you would be something like ``Holder.creationCode`` and ``Holder.runtimeCode`` - read-only properties of type bytes memory. What do you think?

Yes, that would definitely be helpful. And also the code hash, of course, because it can be computed at compile time

**Comment by chriseth** [comment]
The optimizer might take care of improving the hash computation, let's see.

**Comment by chriseth** [comment]
Created an issue https://github.com/ethereum/solidity/issues/5647

**Comment by jochem-brouwer** [comment]
@chriseth Your `bytes memory Holder.runtimeCode` sounds dangerous. It is possible to create contracts which dump different runtime code than what you might expect. This involves using the `return` opcode in the constructor via assembly by hence pushing different bytes to the deployed code than what solidity expects it will deploy.

**Comment by chriseth** [comment]
Let's not discuss in two different issues -> https://github.com/ethereum/solidity/issues/5647

**Comment by MicahZoltu** [comment]
Separate from the `.creationCode` and `.runtimeCode`, I would still like to see a `new2` operator (that takes a salt).  Dropping down to assembly should only happen when the higher level language doesn't support what you need to do, and if people are frequently dropping to assembly (as is happening with `create2`) that suggests the language is missing a feature that is in high demand.

Sometimes people drop into assembly to do things they _shouldn't_ be doing, and I don't think we should be adding Solidity language features for such operations.  However, using `create2` is a good thing and should (IMO) be encouraged, but that requires a Solidity level language construct for it.

**Comment by haydenadams** [comment]
Whats the status of this? Would be nice to have a non-assembly CREATE2

**Comment by chriseth** [comment]
Any there any suggestions for the syntax? I really dislike the `2` in the opcode, by the way. What about

`salted_new(salt) C(a, b, c)`

or just

`new(salt) C(a, b, c)`?

We might even go so far to use

`new(salt, value) C(a, b,c)`

to replace `(new(salt) C).value(value)(a, b,c)`

**Comment by chriseth** [comment]
Suggestion by @leonardoalt : `new.salt(s) C(a, b, c)`

**Comment by chriseth** [comment]
Another suggestion by @leonardoalt: Also change to `new.value(1 ether) C(a, b, c)`

**Comment by leonardoalt** [comment]
And then you can combine them:
`new.value(1000 ether).salt(42) C(a, b, c);`

**Comment by chriseth** [comment]
I like `new.salt(s)` - we could introduce that with a non-breaking release already - although the parser looks a bit involved, so it might break the AST.

**Comment by chriseth** [comment]
We might change it so that `.value` and `.salt` are special expressions directly in the AST, though, instead of generic members of expressions.

**Comment by leonardoalt** [comment]
Suggestion from @axic:
`new{salt=...,value=1} C(..)`
`new<salt=...,value=1> C(..)`
If something like the above is accepted, the idea is to remove `.value` and `.gas` for all functions.

**Comment by chriseth** [comment]
I also like `new{salt=..., value=...} C(...)`, but it's probably harder to parse `<expression>{value=..., gas=...}()`.

**Comment by ekpyron** [comment]
Some options in use in multiple contexts as @axic requested...
```
  c.f<value=42, gas=1>(a, b, c);
  (cond ? c.f : g.h)<value=42, gas=1>(a, b, c);
  new<value=42, salt=...> C(a,b,c);
  new C<value=42, salt=...>(a,b,c);

  c.f{value=42, gas=1}(a, b, c);
  (cond ? c.f : g.h){value=42, gas=1}(a, b, c);
  new{value=42, salt=...} C(a,b,c);
  new C{value=42, salt=...}(a,b,c);

  c.f(a, b, c) [[value=42, gas=1]];
  (cond ? c.f : g.h)(a, b, c) [[value=42, gas=1]];
  new C(a,b,c) [[value=42, salt=...]];

  c.f[[value=42, gas=1]](a, b, c);
  (cond ? c.f : g.h)[[value=42, gas=1]](a, b, c);
  new C[[value=42, salt=...]](a,b,c);
  new[[value=42, salt=...]] C(a,b,c);

```

**Comment by ekpyron** [comment]
```
  { c.f, value=42, gas=1 }(a, b, c);
  { cond ? c.f : g.h, value=42, gas=1}(a, b, c);
  {new, value=42, salt=...} C(a,b,c);
  new {C, value=42, salt=...}(a,b,c);

  < c.f, value=42, gas=1 >(a, b, c);
  < cond ? c.f : g.h, value=42, gas=1>(a, b, c);
  <new, value=42, salt=...> C(a,b,c);
  new <C, value=42, salt=...>(a,b,c);

  [[ c.f, value=42, gas=1 ]] (a, b, c);
  [[ cond ? c.f : g.h, value=42, gas=1]] (a, b, c);
  [[new, value=42, salt=...]] C(a,b,c);
  new [[C, value=42, salt=...]](a,b,c);
```

**Comment by chriseth** [comment]
```
c.f{value=3, gas=8}(a, b)
new C{value=3, salt=8}(a, b)
(cond ? c.f : g.h){value=42, gas=1}(a, b, c);
```

**Comment by MicahZoltu** [comment]
I'm pretty strongly against `C` and `c.f` and `g.h` or anything else with single letter variable names.  Words don't take up meaningful amounts of space, and clarity is king in smart contracts.  We should always prefer verbosity over brevity in any language constructs IMO.

**Comment by chriseth** [comment]
@MicahZoltu these are just examples of contract and function names, they are non-normative :)

**Comment by ekpyron** [comment]
At the moment I like these best:
```
{c.f, value=3, gas=8}(a, b)
{new C, value=3, salt=8}(a, b)
{cond ? c.f : g.h, value=42, gas=1}(a, b, c);
```

**Comment by axic** [comment]
My choice would be:
```
c.f{value=3, gas=8}(a, b)
new C(a, b)
new C{value=3, salt=8}(a, b)
(cond ? c.f : g.h){value=42, gas=1}(a, b, c);
```

**Comment by axic** [comment]
Actually based on discussion with @ekpyron if you consider the curly brackets to "wrap functions with a modifier", then the following makes sen:
```
  { c.f, value=42, gas=1 }(a, b, c);
  { cond ? c.f : g.h, value=42, gas=1}(a, b, c);
  {new C, value=42}(a,b,c);
  {new C, value=42, salt=...}(a,b,c);
```

In this case, the following be valid?
```
  {c.f}();
  {new C}();
```

**Comment by ekpyron** [comment]
```
c.f{value=3, gas=8, (a, b)}
new C{ value=3, salt=8, (a, b)}
(cond ? c.f : g.h){ value=42, gas=1, (a, b, c) };
```

Yeah, probably not...

**Comment by axic** [comment]
> Function create2_address, which allows computing address of CREATE2 contract. I would have used it twice, first in Token.getHolderContract function: address payable holder_address = create2_address(factory, a, Holder), and second in Holder.setOwner function: require(create2_address(factor, _owner, Holder) == address(this))

Since we have `type(Holder).creationCode`, how about adding `type(Holder).calculateSaltedAddress(salt)` ?

**Comment by chriseth** [comment]
`calculatedSaltedAddress` would also need the constructor arguments - so rather
`calculateAddress{salt = ...}(arg1, arg2, ..., argn)`

**Comment by ekpyron** [comment]
Or ``{new Holder, salt= ...}.address(arg1, arg2, ..., argn)``?

**Comment by ekpyron** [comment]
``type(Holder{salt = ... }).calculateAddress(arg1, arg, ... , argn)``?

**Comment by axic** [comment]
Maybe we actually need to look at this in conjunction with #2296 and that we could consider removing the `new` keyword entirely.

**Comment by leonardoalt** [comment]
Folks seem to converge to `{...}`, but I prefer `[[...]]` :p

**Comment by ekpyron** [comment]
> Folks seem to converge to `{...}`, but I prefer `[[...]]` :p

Haha, I actually still like ``[[...]]`` better as well, but dropped it following the general tendency :-).

**Comment by axic** [comment]
``[[..]]`` feels more like an annotation on a function declaration as opposed to an expression.

**Comment by 3esmit** [comment]
I am writing account contracts and there is a specific use case where I am having to override a function because I am unsure about gas operator. 
The override takes place to set if gas is manual or not. 
There is a magic value (such as 0 or MAX_INT)? In case of `{gas = 0}`, what happens? and in case `{gas = MAX_INT}` (MAX_INT can be any value greater then the block gas limit)? To set the gas to "auto" I have to use `{gas = gasleft()}` ?

**Comment by chriseth** [comment]
Decsion: We use `salt=` as a determinator to use create / create2.

**Comment by chriseth** [comment]
Decision: We use `{salt=...}` and also go with `{value=...,gas=...}` for all external function calls.

**Comment by axic** [comment]
But is it `new{} C` or `new C{}` ?

**Comment by chriseth** [comment]
`new C{}()`

**Comment by chriseth** [comment]
Implemented in #8177
